CSV export (example implementation in user list)

MaximilianKrambach opened the pull request:
This does the csv export for the user list

The csvExportService is generic and can be used from other places as well (see user-list.component.ts: csvExportUserList)

I use the `file-saver` package, which works well. I decided against the `angular-file-saver` package, which is only an aditional wrapper around file-saver and Blob. If we need to support a browser that knows nothing about Blob, we might switch to angular-file-saver dependency instead.

I did not care about permission checks prior to exporting, as my main focus was on the export itself, and the user export is just the random example I picked.

MaximilianKrambach commented:
ping @FinnStutzenstein for review

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,118 @@
+     * @param filename name of the resulting file
+     * @param options
+     */
+    public saveAsCsv(
```
`public save<T extends BaseViewModel>(data: T[], columns: { [P in keyof T]?/* the property*/: string /* the column title*/}, filename: string, ...rest)`
Does it work to make `{ineSeparator = '\r\n', columnSeparator = ';'}` optional. If you do not want to provide values, you have to pass an empty object; this is not nice.

MaximilianKrambach replied:
```
{ lineSeparator = '\r\n', columnSeparator = ';' }:{ lineSeparator?: string; columnSeparator?: string } = {}
```
This should work (at least in es2016 this is how destructuring works with an default of 'no data submitted'; I'm testing it now)

FinnStutzenstein replied:
Perfect!

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,118 @@
+
+        // create header data
+        const header = [];
+        for (const column of columns) {
```
Here you have to interate through the key-values in the columns object.

MaximilianKrambach replied:
So, we'll lose the 'label for this column is optional'. If that's fine, I'll continue with this

FinnStutzenstein replied:
yeah. If we have too much of the same names, you can define the object values as `string|null`, null meaning "take the property value".

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,118 @@
+        for (const column of columns) {
+            if (column.title) {
+                ts_list = this.checkCsvTextSafety(column.title, ts_list);
+                header.push(column.title);
```
Should this be translated?

MaximilianKrambach replied:
I would not translate it during csv export, but rather on the other end (the listView), where a translationService is already supposed to be in place

FinnStutzenstein replied:
You can just put the `TranslateService` into the constructor. Then we have one central place where the values get translated and the views do not have to care about this (should be less code there then).

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,118 @@
+        // assemble lines, putting text separator in place
+
+        const final_lines = [];
+        if (ts_list.length) {
```
swap this to reduce indentions:
```
if (!ts_list.length) {
    throw ...
}
...
```

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,118 @@
+                final_lines.push(newline.join(columnSeparator));
+            }
+            this.exporter.saveFile(final_lines.join(lineSeparator), filename);
+            return true;
```
remove this and make the function return void. If somethings happens, an error can be thrown.

FinnStutzenstein commented on client/src/app/core/services/file-export.service.ts:
```diff
@@ -0,0 +1,45 @@
+     * @param file
+     * @param filename
+     * TODO: Potential encoding issues and path/filename oddities on different browsers
+     * also: default filename
```
I would suggest making the filename not optional.

FinnStutzenstein commented on client/src/app/core/services/file-export.service.ts:
```diff
@@ -0,0 +1,45 @@
+    protected validateFileName(filename: string): boolean {
+        // everything not containing \/?%*:|"<> and not ending on a dot
+        const pattern = new RegExp(/^[^\\\/\?%\*:\|\"\<\>]*[^\.]+$/i);
+        if (pattern.test(filename)) {
```
just do `return pattern.test(filename);`.

FinnStutzenstein commented on client/src/app/site/users/components/user-list/user-list.component.ts:
```diff
@@ -99,4 +101,18 @@ export class UserListComponent extends ListViewBaseComponent<ViewUser> implement
+    public saveUserList(): void {
+        this.csvExport.saveAsCsv(
+            this.dataSource.data,
+            [
```
```
{
    fullName: 'Name',
    groups: 'Groups',
    isActive: 'Presence'
}
```

MaximilianKrambach replied:
(direct discussion: we keep it an array of objects: Order will be preserved, and label can be optional (capitalized version of the property will be translated and used, if no label present)

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -109,4 +116,13 @@ export class CsvExportService {
         return ts_list;
     }
+
+    private capitalizeTranslate(input: any): string {
```
`input: string` and remove the check below.

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,128 @@
+        // create header data
+        const header = [];
+        columns.forEach(column => {
+            const rawLabel = column.label ? column.label : column.property;
```
then declare `rawLabel` as string and cast the property: `column.property as string`.

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,124 @@
+import { TranslateService } from '@ngx-translate/core';
+
+/**
+ * Handles all incoming and outgoing notify messages via {@link WebsocketService}.
```
Isn't this a bit off? ;)

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,124 @@
+     *      lineSeparator (defaults to \r\n windows style line separator),
+     *      columnseparator defaults to semicolon (other usual separators are ',' '\T' (tab), ' 'whitespace)
+     */
+    public saveAsCsv<T extends BaseViewModel>(
```
Just name it `save` or `export`.

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,124 @@
+        filename: string,
+        { lineSeparator = '\r\n', columnSeparator = ';' }: { lineSeparator?: string; columnSeparator?: string } = {}
+    ): void {
+        const all_lines = []; // Array of arrays of entries
```
Please make all variables camelCase

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,124 @@
+        all_lines.push(header);
+        // create lines
+        data.forEach(item => {
+            const line = [];
```
```
 const line = used_columns.map(property => {
    ts_list = this.checkCsvTextSafety(item[property], ts_list);
    return item[property];
});
```

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,124 @@
+        if (!ts_list.length) {
+            throw new Error('no usable text separator left for valid csv text');
+        }
+        all_lines.forEach(function(line: any[]): void {
```
`const finalLines = allLines.map(line => {... return newline.join(columnSeparator)});`

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,124 @@
+        }
+        all_lines.forEach(function(line: any[]): void {
+            const newline = [];
+            line.forEach(function(item: any): void {
```
`line.forEach(item => {`

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,124 @@
+     * @param input any input to be sent to CSV
+     * @param ts_list The list of special characters to check.
+     */
+    public checkCsvTextSafety(input: any, ts_list: string[]): string[] {
```
type of input: `string`. Cast above if accessing `item[property]`.

The complete function may be cut down to `return tsList.filter(char => input.indexOf(char) < 0);`. This needs testing but should work.

MaximilianKrambach replied:
At this type I should not be required to rely on it being a string: For later parsing, I will need to escape texts into a textseparator (which I'm finding out at this stage). If I cast is to a string above, I'll lose the ability to check for numbers or booleans further in the line assembly

emanuelschuetze commented on client/src/app/site/users/components/user-list/user-list.component.ts:
```diff
@@ -69,4 +93,17 @@ export class UserListComponent extends ListViewBaseComponent<ViewUser> implement
     }
+
+    // TODO save all data from the dataSource
+    public saveUserList(): void {
```
Please use 'csv' in the function name, e.g. `csvExportUserList()`, because later we'll export the user list also in PDF. See function names in 2.3.x (`pdfExportUserList()`, `pdfExportUserAccessDataList()`, `csvExport())`: https://github.com/OpenSlides/OpenSlides/blob/stable/2.3.x/openslides/users/static/templates/users/user-list.html#L67

Also the menu button should show 'CSV', e.g. 'Export (CSV)'.

emanuelschuetze commented on client/src/app/site/users/components/user-list/user-list.component.ts:
```diff
@@ -69,4 +93,17 @@ export class UserListComponent extends ListViewBaseComponent<ViewUser> implement
+            [
+                { property: 'full_name', label: 'Name' },
+                { property: 'groups', label: 'Groups' },
+                { property: 'is_active', label: 'Presence' }
```
Please use the same csv columns as in 2.3.x:
https://github.com/OpenSlides/OpenSlides/blob/stable/2.3.x/openslides/users/static/js/users/csv.js#L14

MaximilianKrambach replied:
Excepting 'number' (which is not part of the data I can access in the user list), I included all data now

FinnStutzenstein replied:
it's named `participant_number`.

emanuelschuetze commented:
@MaximilianKrambach This pull requst ist for user csv export only, right? Should be mentioned in commit message / pr title. Still open: agenda and motions csv export.

MaximilianKrambach commented:
I updated according to your comments

ping @FinnStutzenstein @emanuelschuetze

MaximilianKrambach commented:
- Okay, a new type of menu was sneaked in; fixed.

- The other problem is that an openslide-specific overload destroyed `Array.join()`. I commented out  the overload, and we need to talk about how to deal with it (ping @FinnStutzenstein ).

- The other hiccup I stumbled on is that user.groups is not a string or number, but an array of sub-items. I think I fixed handling there, but it may not be clear what to do.

FinnStutzenstein commented on client/src/app/app.component.ts:
```diff
@@ -54,22 +54,24 @@ export class AppComponent {
      * TODO: Overloading can be extended to more functions.
      */
     private overloadArrayToString(): void {
-        Array.prototype.toString = function(): string {
```
I checked and this was not a problem to comment in. Please verify

MaximilianKrambach replied:
Okay, with your (slightly different) ways of assembling a list we indeed avoid this problem; I am still unhappy with sneakily overwriting natiive functions.

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -0,0 +1,123 @@
+                    item[property].forEach(subitem => {
+                        subproperties.push(subitem[columns[i].assemble]);
+                    });
+                    prop = subproperties.join(' ');
```
`prop = item[property].map(subitem => this.translate.instant(subitem[columns[i].assemble])).join(',');`
Also remove the `,` from tsList.

MaximilianKrambach replied:
So, the assumptions are: The data of the assembled subitems are always strings to be translated, and they will be comma separated. Well then.

FinnStutzenstein replied:
yes. Maybe ad a toString then; but we can change this, if it is the case.

MaximilianKrambach commented:
I'll rebase tomorrow.

MaximilianKrambach commented:
@FinnStutzenstein your remarks have been addressed as desired, and I rebased and squished the commits. Mergeable in my opinion
